### DISCUSSION TITLE
Record audio generate by my app, not the mic

### DISCUSSION THREAD

**Discussion by paulorrfilho** [google/oboe#1258] [Q&A] [answered]
I understand this is likely out of scope but here is my question:

My app allows users to play an instrument along with mp3 audio files. The instrument is powered by Oboe but the mp3 files are played using a MediaPlayer. I want to record/capture the audio generated by my app, not the mic, and save it to an mp3 or other compressed format file.

Possible solutions:

1) Play the mp3 files also using Oboe and save the stream data to an mp3 file. I believe this is possible but hard to implement.

2) Somehow record my app's audio output. Lazy way. I could not find a way to achieve this, am I missing something?

Can you please advise on this?

**Comment by dturner** [comment]
> Play the mp3 files also using Oboe and save the stream data to an mp3 file. I believe this is possible but hard to implement.

Certainly possible, but yes, more difficult to implement than `MediaPlayer`. That said, there is a code sample which decodes MP3s and plays them here https://github.com/google/oboe/tree/master/samples/RhythmGame which you could use as a starting point. This is probably the correct way of doing it. 

>  Somehow record my app's audio output. Lazy way. I could not find a way to achieve this, am I missing something?

You could look at using https://developer.android.com/guide/topics/media/playback-capture, however I'm not certain that it works for apps to record their own audio. Should be quite simple to test out though.

**Reply by paulorrfilho** [reply]
Thank you @dturner, I looked into Playback Capture in the past but it does not seem to support older phones. I think I will go with the most difficult path, I am sure it will pay off in the end.

I am just struggling to understand how to store the stream data to be saved later. The Oboe Tester app has a recording capability but it is for the microphone stream. Also, there is so much overhead in the code that it makes it hard to understand what is actually needed. I suggest creating both "hello-oboe-recording-stream" and "hello-oboe-recording-mic" sample apps.

**Comment by philburk** [comment, marked as answer]
If you want capture the output of your app then this may work.

Create an atomic FifoBuffer using the internal FIFO tool from Oboe.
https://github.com/google/oboe/blob/master/src/fifo/FifoBuffer.h
Make it fairly large, like a half second worth of data.

uint32_t capacity = sampleRate / 2;
FifoBuffer *fifo = new FifoBuffer(stream->getBytesPerFrame(), capacity);

Launch a thread that will periodically, maybe every 50 msec, will wake up and 
read() from the FIFO and write the data to a file.
The fifo->read() method will return the number of frames actually read.

Then in your callback, write() a copy of your audio data to the FIFO.

It is important that the file is written in a separate thread, not the callback, because file I/O can block and cause glitches in the audio.
